Test type_caster bare interface

rwgk opened the pull request:
Systematically exercises the `detail::type_caster<> interface` (which is a de-facto public interface). This is going a step in the direction of an integration test, to ensure multiple components of pybind11 work together correctly. It is also useful to show the `type_caster<>` interface virtually clutter-free.

This new test was copied from PR #2672, with small modifications.

YannickJadoul commented on tests/test_type_caster_bare_interface.cpp:
```diff
@@ -0,0 +1,215 @@
+namespace pybind11_tests {
+namespace type_caster_bare_interface {
+
+struct mpty {};
```
I have a hard time reading everything since 1) I have no idea how to pronounce/read this, 2) my brain's expecting capitals for "custom"/user-side classes in the test. Do you mind just turning (or I can also do it, if you don't have time) this into `Foo` or `CustomClass` or `Value` or so?

YannickJadoul commented on tests/test_type_caster_bare_interface.cpp:
```diff
@@ -0,0 +1,215 @@
+std::unique_ptr<mpty>       rtrn_uqmp() { return std::unique_ptr<mpty      >(new mpty); }
+std::unique_ptr<mpty const> rtrn_uqcp() { return std::unique_ptr<mpty const>(new mpty); }
+
+const char* pass_uqmp(std::unique_ptr<mpty>)       { return "load_uqmp"; }
```
Uuumpf, this takes me 2 second per function to parse to know what it's saying. Do you feel strongly about keeping 4 letters? Personally, I find `pass_unique_ptr`, `pass_const_ref`, `pass_rvalue_ref`, `return_...`, ... easier to directly read, even though things then don't nicely line out anymore.

But your effort on writing tests, so your final say.

YannickJadoul commented on tests/test_type_caster_bare_interface.cpp:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    template <typename T_>
+    using cast_op_type = conditional_t<
+        std::is_same<remove_reference_t<T_>, mpty const *>::value,
+        mpty const *,
+        conditional_t<
+            std::is_same<remove_reference_t<T_>, mpty *>::value,
+            mpty *,
+            conditional_t<
+                std::is_same<T_, mpty const &>::value,
+                mpty const &,
+                conditional_t<std::is_same<T_, mpty &>::value,
+                              mpty &,
+                              conditional_t<std::is_same<T_, mpty &&>::value, mpty &&, mpty>>>>>;
```
What about going `else if` style on this, to avoid indentation? (similar to [test_builtin_casters.cpp](https://github.com/pybind/pybind11/blob/0432ae7c5222ecb5d6e81babf996d3777bd94bf3/tests/test_builtin_casters.cpp#L42))

```suggestion
    using cast_op_type =
        conditional_t<std::is_same<remove_reference_t<T_>, mpty const *>::value,
            mpty const *,
        conditional_t<std::is_same<remove_reference_t<T_>, mpty *>::value,
            mpty *,
        conditional_t<std::is_same<T_, mpty const &>::value,
            mpty const &,
        conditional_t<std::is_same<T_, mpty &>::value,
            mpty &,
        conditional_t<std::is_same<T_, mpty &&>::value,
            mpty &&,
            /* else */ mpty>>>>>;
```

rhaschke replied:
Isn't all this just returning the same type as detected?

YannickJadoul commented on tests/CMakeLists.txt:
```diff
@@ -129,6 +129,7 @@ set(PYBIND11_TEST_FILES
     test_stl.cpp
     test_stl_binders.cpp
     test_tagbased_polymorphic.cpp
+    test_type_caster_bare_interface.cpp
```
Final nitpick: `test_type_caster_interface` also works, if we want to have this file show off the interface?

YannickJadoul reviewed (commented):
Mostly some style nitpicking, in order to be consistent with the style in other pybind11 tests, and be more intuitively readable. Apart from that, this seems useful to have tested, yes, so thanks a lot for adding this long overdue test!

One content-wise thing: this seems to be a superset of `type_caster<ConstRefCasted>` in `test_const_ref_caster`, except for the `reference_wrapper` part. Should we remove that test, and just copy the `return_reference_wrapper`/`pass_reference_wrapper` function?

Bonus nitpick: pybind11 (reasonably) consistently uses west-side `const`,  `const Foo &bar` style. I realize there's a whole discussion on which one is best and I won't take sides (though I happen to agree with pybind11), but mixing styles might be even worse?

rwgk commented:
Converting this to a Draft PR for now. While developing PR #2672 I find myself tweaking the systematic naming scheme. I will update this PR with matching names later.

rhaschke commented on tests/test_type_caster_bare_interface.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def test_load():
+    assert m.pass_valu(None) == "load_valu"
+    assert m.pass_rref(None) == "load_rref"
+    assert m.pass_cref(None) == "load_cref"
+    assert m.pass_mref(None) == "load_mref"
+    assert m.pass_cptr(None) == "load_cptr"
+    assert m.pass_mptr(None) == "load_mptr"
```
What is actually tested here? This is just about calling different functions and comparing their return values, isn't it?

rhaschke commented on tests/test_type_caster_bare_interface.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def test_load_shared_ptr():
+    assert m.pass_shmp(None) == "load_shmp"
+    assert m.pass_shcp(None) == "load_shcp"
```
Same here.

rhaschke commented on tests/test_type_caster_bare_interface.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def test_load_unique_ptr():
+    assert m.pass_uqmp(None) == "load_uqmp"
+    assert m.pass_uqcp(None) == "load_uqcp"
```
and here.

rhaschke reviewed (commented):
I think, testing that `rtrn_*()` calls the correct casting operation is fine.
However, changing the function signature of `pass_*()` will still call the "old" functions. 
I don't see a value in testing this. Rather, it should be tested whether the values are correctly copied/moved when passing from python to C++.

rwgk commented:
Hi Robert, I reworked this PR quite a bit, to be less the stepping stone it was for me back in December, and a more meaningful test / demo (and be more similar to what's now test_class_sh_basic on the smart_holder branch). Could you please take another look? I'm now tracking `CpCtor` and `MvCtor`. I also removed all the smart-pointer code, because it's not relevant to the copy/move logic we're concerned about here.

The valgrind CI failure seems unrelated, btw; something todo with pypocketfft.

rwgk commented:
Closing this stale PR to reduce distractions.

Files archived here as-is:

https://github.com/rwgk/rwgk_tbx/commit/faec6cbbdfdf9f253911838e29f05585a79867ef
